dadb/src/main/kotlin/dadb/forwarding/TcpForwarder.kt

```diff
@@ -0,0 +1,136 @@
+package dadb.forwarding
+
+import dadb.Dadb
+import okio.BufferedSink
+import okio.Source
+import okio.buffer
+import okio.sink
+import okio.source
+import java.io.IOException
+import java.io.InterruptedIOException
+import java.net.ServerSocket
+import java.util.concurrent.TimeoutException
+import kotlin.concurrent.thread
+
+internal class TcpForwarder(
+    private val dadb: Dadb,
+    private val hostPort: Int,
+    private val targetPort: Int,
+) : AutoCloseable {
+
+    private var state: State = State.STOPPED
+    private var thread: Thread? = null
+
+    fun start() {
+        check(state == State.STOPPED) { "Forwarder is already started at port $hostPort" }
+
+        moveToState(State.STARTING)
+        thread = thread {
+            try {
+                handleForwarding()
+            } finally {
+                moveToState(State.STOPPED)
+            }
+        }
+
+        waitFor(10, 5000) {
```

Comment:
Is this necessary? The process will stay alive since the thread has started at this point already.

Reply:
There is no guarantee that the thread have already started and server is ready to accept connections